### PR TITLE
privilege, session, server: consistently map user login to identity

### PR DESCRIPTION
**Pull request by morgo**
### What problem does this PR solve?

Problem Summary:

TiDB server uses different rules to map a user to an identity depending on the code area. This cleans up how identity is mapped using a consistent function: `MatchIdentity()`.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [x] Breaking backward compatibility
The rules might be slightly different in the event that there are two 'root' users with different hosts (not a common practice, but it's possible users are doing this). This PR follows more consistent behavior with MySQL.

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
The TiDB server now maps a user to an entry in the mysql.user table more consistently.
```

### CODE REVIEW

**Review by dveeden — APPROVED** [review]

**Comment by morgo** [comment]
It looks like this might make one of the tests unstable:
```
[2021-11-29T17:18:37.901Z] [2021/11/30 01:14:48.400 +08:00] [ERROR] [http_status.go:460] ["grpc server error"] [error="mux: listener closed"]
[2021-11-29T17:18:37.901Z] --- FAIL: TestErrorCode (0.74s)
[2021-11-29T17:18:37.901Z]     server_test.go:137: 
[2021-11-29T17:18:37.901Z]         	Error Trace:	server_test.go:137
[2021-11-29T17:18:37.901Z]         	            				server_test.go:1427
[2021-11-29T17:18:37.901Z]         	            				tidb_test.go:194
[2021-11-29T17:18:37.901Z]         	Error:      	Received unexpected error:
[2021-11-29T17:18:37.901Z]         	            	No Pem data found, data: #28000Access denied for user 'root'@'127.0.0.1' (using password: YES)
[2021-11-29T17:18:37.901Z]         	Test:       	TestErrorCode
[2021-11-29T17:18:37.901Z]         	Messages:   	Error drop database ErrorCode: No Pem data found, data: #28000Access denied for user 'root'@'127.0.0.1' (using password: YES)
```

I can't tell if it is related yet.

File: privilege/privileges/cache.go [hunk 1]
```diff
@@ -902,14 +906,54 @@ func patternMatch(str string, patChars, patTypes []byte) bool {
...
 		if record.match(user, host) {
 			return record
 		}
 	}
+
+	// If skip-name resolve is not enabled, and the host is not localhost
+	// we can fallback and try to resolve with all addrs that match.
+	// TODO: this is imported from previous code in session.Auth(), and can be improved in future.
+	if !skipNameResolve && host != variable.DefHostname {
+		addrs, err := net.LookupAddr(host)
```
Comment (tiancaiamao) [inline privilege/privileges/cache.go:923]: This code branch may have some performance impact if `matchIdentity` is frequently called.

Comment (tiancaiamao) [reply privilege/privileges/cache.go:923]: If it's just used in connection verification, the performance impact is OK
But if it's used in query privilege verification, I'm afraid the impact is not negligible

**Review by tiancaiamao — APPROVED** [review]

**Pull request merged**